### PR TITLE
[Core] Added UNIQUE constraint to Subtest_name in instrument_subtests

### PR DESCRIPTION
**Pull request by ZainVirani**
Ensures the front end only shows one link to each page of an instrument via the UNIQUE constraint in instrument_subtests.

### CODE REVIEW

File: SQL/0000-00-00-schema.sql [hunk 1]
```diff
@@ -359,6 +359,7 @@ CREATE TABLE `instrument_subtests` (
   `Subtest_name` varchar(255) NOT NULL default '',
   `Description` varchar(255) NOT NULL default '',
   `Order_number` int(11) NOT NULL default '0',
+  UNIQUE (Subtest_name),
```
Comment (xlecours) [inline SQL/0000-00-00-schema.sql]: I would suggest to explicitly name the unique key.
ex:
 ```
UNIQUE KEY `Subtest_name` (`Subtest_name`)
```

File: SQL/Archive/17.1/2017-05-11-unique-sub-tests.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+WARNINGS;
+SET SQL_NOTES=1;
```
Comment (ridz1208) [inline SQL/Archive/17.1/2017-05-11-unique-sub-tests.sql]: @xlecours should the original value be saved to a variable first and reset after the patch ?

**Comment by ridz1208** [comment]
@ZainVirani What happens if duplicates do exist in the table ?

**Comment by ZainVirani** [comment]
@ridz1208 Under Subtests on the left sidebar multiple links to the same page of an instrument can appear (Page 1, Page 1, Page 1, all of which link to page 1 of the instrument). This only occurs if an INSERT statement to instrument_subtests with the same values is run more than once.

**Comment by ridz1208** [comment]
@ZainVirani and if you run this patch what will happen to the non-unique values ? is it gonna throw an error ??

**Comment by ZainVirani** [comment]
@ridz1208 the patch fails
ERROR 1062 (23000): Duplicate entry 'VALUE' for key 'Subtest_name'

**Comment by ZainVirani** [comment]
@ridz1208 edited the sql patch to delete duplicate entries before adding the UNIQUE constraint for key Subtest_name

File: SQL/0000-00-00-schema.sql [hunk 3]
```diff
@@ -359,6 +359,7 @@ CREATE TABLE `instrument_subtests` (
   `Subtest_name` varchar(255) NOT NULL default '',
   `Description` varchar(255) NOT NULL default '',
   `Order_number` int(11) NOT NULL default '0',
+  UNIQUE KEY `Subtest_name` (`Subtest_name`),
```
Comment (driusan) [inline SQL/0000-00-00-schema.sql]: I don't think this constraint makes sense. Subtest_name is the name of the page, but different instruments might have pages with the same name. (ie. different instruments might both have different pages both named "page2")

The constraint should be on the combination of `(Test_name, Subtest_name)`

**Comment by driusan** [comment]
@xlecours can you dismiss your review or re-review?

File: SQL/Archive/17.1/2017-05-11-unique-sub-tests.sql (hunk 2 above)
Comment (xlecours) [inline SQL/Archive/17.1/2017-05-11-unique-sub-tests.sql]: @ridz1208 No, it is a session variable.

File: SQL/Archive/17.1/2017-05-11-unique-sub-tests.sql [hunk 4]
```diff
@@ -1,7 +1,25 @@
 WARNINGS;
 SET SQL_NOTES=1;
 
-SELECT 'Alter instrument_subtests to force unique Subtest_name' as 'Step #1';
+SELECT 'Delete duplicate entries' as 'Step #1';
+DELETE
```
Comment (xlecours) [inline SQL/Archive/17.1/2017-05-11-unique-sub-tests.sql]: ERROR 1055 (42000): Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'LORIS_17_1.instrument_subtests.ID' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by

I would suggest usign this synthax:
`
DELETE FROM instrument_subtests USING instrument_subtests, instrument_subtests to_keep
  WHERE instrument_subtests.ID < to_keep.ID AND instrument_subtests.Test_name = to_keep.Test_name AND instrument_subtests.Subtest_name = to_keep.Subtest_name;
`
This should keep the latest record according to auto_increment.

**Comment by ZainVirani** [comment]
@xlecours changes have been made

**Comment by sruthymathew123** [comment]
The script ran smooth on my database. Seems it added the unique key constaint on the defined fields. After that I tried creating a duplicate entry and the duplicate entry warning was poping up.
So I am putting passed manual tests from my part.

**Comment by ridz1208** [comment]
@davidblader please test as well on our DB

**Comment by davidblader** [comment]
LGTM

**Pull request merged**